pkg/agent/plugin/workloadattestor/k8s/k8s.go

```diff
@@ -93,59 +102,115 @@ func (p *k8sPlugin) Attest(req *workloadattestor.AttestRequest) (*workloadattest
 		return &resp, nil
 	}
 
-	httpResp, err := p.httpClient.Get(fmt.Sprintf("http://localhost:%v/pods", p.kubeletReadOnlyPort))
-	if err != nil {
-		return &resp, err
+	// Poll pod information and search for the pod with the container. If
+	// the pod is not found, and there are pods with containers that aren't
+	// fully initialized, delay for a little bit and try again.
+	for attempt := 1; ; attempt++ {
+		podInfo, err := p.getPodInfoFromInsecureKubeletPort()
+		if err != nil {
+			return &resp, err
+		}
+
+		notAllContainersReady := false
+		for _, item := range podInfo.Items {
+			switch lookUpContainerInPod(containerID, item.Status) {
+			case containerInPod:
+				resp.Selectors = append(resp.Selectors, &common.Selector{Type: selectorType, Value: fmt.Sprintf("sa:%v", item.Spec.ServiceAccountName)})
+				resp.Selectors = append(resp.Selectors, &common.Selector{Type: selectorType, Value: fmt.Sprintf("ns:%v", item.Metadata.Namespace)})
+				return &resp, nil
+			case containerMaybeInPod:
+				notAllContainersReady = true
+			case containerNotInPod:
+			}
+		}
+
+		// if the container was not located and there were no pods with
+		// uninitialized containers, then the search is over.
+		if !notAllContainersReady || attempt >= p.maxPollAttempts {
+			log.Printf("container id %q not found (attempt %d of %d)", containerID, attempt, p.maxPollAttempts)
+			return &resp, fmt.Errorf("no selectors found")
+		}
+
+		// wait a bit for containers to initialize before trying again.
+		log.Printf("container id %q not found (attempt %d of %d); trying again in %s", containerID, attempt, p.maxPollAttempts, p.pollRetryInterval)
+
+		// TODO: bail early via context cancelation
```

Comment:
i'd love if we could thread through contexts into the plugins at some point to support cancelation(amongst other things)